### ISSUE TITLE
2.5.19 introduced kind mismatch error into another program

### ISSUE THREAD

**Issue by temoto** [ndmitchell/derive#5]
Please see https://github.com/Peaker/lamdu/issues/27
code like 
```
{-# LANGUAGE KindSignatures, TemplateHaskell, DeriveFunctor, DeriveFoldable, DeriveTraversable, GeneralizedNewtypeDeriving, DeriveDataTypeable, RankNTypes #-}
-- hopefully didn't cut something important to understanding
Lens.makeLenses ''Scope
derive makeMonoid ''Scope
```
produces kind mismatch error using derive 2.5.19
```
Lamdu/Sugar/Types.hs:491:1:
    The second argument of ‘Scope’ should have kind ‘* -> *’,
      but ‘m_1628154822’ has kind ‘*’
    In the instance declaration for
      ‘Monoid (Scope name_1628154821 m_1628154822)’
```
but works using 2.5.18.

**Comment by ndmitchell** [comment]
Sorry it's taken a little while to get to this, especially as it's a regression. My computer suddenly died about the time you reported this error, and I'm only really now back up and running properly. A minimal example, avoiding Lamdu, is:

    import Data.Derive.Monoid
    import Data.DeriveTH
    import Data.Monoid

    data Foo a b = Foo (a b)

    data Scope name m = Scope
      { _scopeLocals    :: [Foo m name]
      , _scopeGlobals   :: [Foo m name]
      }

    derive makeMonoid ''Scope

**Comment by ndmitchell** [comment]
The issue was that in GHC 7.8 and below, the only type variables that got kind annotations were those which weren't kind `*`, and due to accidental luck, I stripped those from contexts. In GHC 7.10 all variables got kind annotations, so I stripped them all on all versions, and that broke the example above as `m` is kind `* -> *`. I fixed that by slightly more intelligent handling of kind annotations, treating no annotation and an annotation of `*` as equivalent. I've just released derive-2.5.20 with the fix.
